pydra/engine/submitter.py

```diff
@@ -33,152 +29,108 @@ def __enter__(self):
     def __exit__(self, type, value, traceback):
         self.close()
 
-    def run(self, runnable, cache_locations=None):
-        """main running method, checks if submitter id for Node or Workflow"""
-        if not isinstance(runnable, TaskBase):  # a node/workflow
-            raise Exception("runnable has to be a Node or Workflow")
-        if runnable.state:
-            runnable.state.prepare_states(runnable.inputs)
-            runnable.state.prepare_inputs()
-        futures = []
-        if runnable.state:
-            for ii, ind in enumerate(runnable.state.states_val):
-                # creating a taskFunction for every element of state
-                # this job will run interface (and will not have state)
-                job = runnable.to_job(ii)
-                checksum = job.checksum
-                # run method has to have checksum to check the existing results
-                job.results_dict[None] = (None, checksum)
-                if cache_locations:
-                    job.cache_locations = cache_locations
-                res = self.worker.run_el(job)
-                futures.append([ii, res, checksum])
-        else:
-            job = runnable.to_job(None)
+    async def fetch_pending(self, wf):
+        """
+        Await worker for any finished tasks.
+        If any task is a workflow, return it.
+        """
+        done = await self.worker.fetch_finished()
+        for fut in done:
+            task = await fut
+            self.worker._remove_pending(fut)
+            logger.debug("Task: %s completed", task)
+            if is_workflow(task):
+                wf.name2obj.get(task.name).__dict__.update(task.__dict__)
```

Comment:
@djarecka because a workflow will expand and execute in a different thread, memory is not shared - this part ensures the object is updated in the preceding thread.

Reply:
@mgxd - since a workflow is a task, the only task that the preceding thread should care about is the workflow task, not its internals. as long as the workflow task returns a future, shouldn't that be sufficient?

Reply:
@mgxd - not sure if I asked about this part yet...

Reply:
ideally, but without this `update` call, the inner workflow's task results are not shared between threads

Reply:
why would they need to be shared between threads? since we can read results from disk anytime.

Reply:
that sounds good to me - we would still have to load the results, check if the inputs have changed, and then assign to the downstream task

Reply:
doesn't each task do this on its own? the only place we were assigning things was at the beginning of a workflow task. when we called `inputs.retrieve_values.`

Reply:
just to clarify: if this `update` is excluded, the inner workflow result will still have a `result()`. However, the tasks within the inner workflow will not, and execution will be stuck (inner workflow will not be marked as `done` since its tasks have no `result()`).